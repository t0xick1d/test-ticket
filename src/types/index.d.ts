// declare module '*.module.scss' {
//    const classes: { [key: string]: string };
//    const content: React.FunctionComponent<React.SVGAttributes<SVGElement>>;
//    export default classes;
// }
