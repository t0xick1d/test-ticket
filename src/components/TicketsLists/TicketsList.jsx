import React from 'react';
import Item from './Item';
import style from './style.module.scss';
import { Grid, Box, Card } from '@mui/material';

const TicketsList = ({ list }) => {
   return (
      // <Grid className={style.containerList}>
      <Box sx={{ flexGrow: 1 }}>
         <Grid
            container
            spacing={4}
            sx={{
               justifyContent: 'center',
               alignItems: 'center',
            }}
            columnSpacing={{ xs: 1, sm: 2, md: 1 }}
            className={style.containerList}
         >
            {list ? list.map((e, i) => <Card key={`${i}list`} children={<Item data={e} />} />) : ''}
         </Grid>
      </Box>
   );
};

export default TicketsList;
