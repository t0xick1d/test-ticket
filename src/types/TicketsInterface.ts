export interface TicketsI {
   price: number;
   carrier: number;
   date: string;
   segments: [SegmentsI];
}
export interface SegmentsI {
   origin: string;
   destination: string;
   departureTime: string;
   duration: number;
   stops: string[];
}
