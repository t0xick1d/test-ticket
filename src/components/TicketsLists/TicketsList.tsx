import React from 'react';
import Item from './Item';
import style from './style.module.scss';
import { Grid, Box, Card } from '@mui/material';
import { TicketsI, TransplantsI } from '../../types/TicketsInterface';
import { filterTransplants } from '../../utils/ticketsUtils/ticketsUtils';
interface MyComponentProps {
   list: TicketsI[];
   transplants: TransplantsI;
   alignment: string;
}

const TicketsList: React.FC<MyComponentProps> = ({ list, transplants, alignment }) => {
   const itemsCopy = list.filter((e) => filterTransplants(transplants, e));
   if (alignment === 'cheap') {
      itemsCopy.sort((a: TicketsI, b: TicketsI) => a.price - b.price);
   }
   if (alignment === 'speed') {
      itemsCopy.sort(
         (a, b) =>
            a.segments.reduce((acc, e) => acc + e.duration, 0) -
            b.segments.reduce((acc, e) => acc + e.duration, 0),
      );
   }
   return (
      <Box sx={{ flexGrow: 1 }}>
         <Grid
            container
            spacing={4}
            sx={{
               justifyContent: 'center',
               alignItems: 'center',
            }}
            columnSpacing={{ xs: 1, sm: 2, md: 1 }}
            className={style.containerList}
         >
            {itemsCopy
               ? itemsCopy.map((e, i) => <Card key={`${i}list`} children={<Item data={e} />} />)
               : ''}
         </Grid>
      </Box>
   );
};

export default TicketsList;
