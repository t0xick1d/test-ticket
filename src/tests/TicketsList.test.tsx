import { render, screen } from '@testing-library/react';
import TicketsList from '../components/TicketsLists/TicketsList';
import { TicketsI, TransplantsI } from '../types/TicketsInterface';

const mockTickets: TicketsI[] = [
   {
      price: 10000,
      carrier: 'SU',
      date: '2025-06-01',
      segments: [
         {
            origin: 'MOW',
            destination: 'HKT',
            departureTime: '2025-06-17T10:30:00',
            duration: 180,
            stops: [],
         },
      ],
   },
   {
      price: 15000,
      carrier: 'S7',
      date: '2025-06-01',
      segments: [
         {
            origin: 'MOW',
            destination: 'HKT',
            departureTime: '2025-06-17T11:00:00',
            duration: 120,
            stops: ['DXB'],
         },
      ],
   },
   {
      price: 12000,
      carrier: 'S7',
      date: '2025-06-01',
      segments: [
         {
            origin: 'MOW',
            destination: 'HKT',
            departureTime: '2025-06-17T11:00:00',
            duration: 120,
            stops: ['DXB'],
         },
      ],
   },
];

const baseTransplants: TransplantsI = {
   all: false,
   noneTransplants: false,
   oneTransplants: false,
   twoTransplants: false,
   threeTransplants: false,
};

describe('TicketsList component', () => {
   it('should check first tickets by price when alignment is "cheap"', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="cheap"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(screen.getByText(/10000 P/i)).toHaveTextContent('10000 P');
   });
   it('should check second tickets by price when alignment is "cheap"', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="cheap"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(screen.getByText(/12000 P/i)).toHaveTextContent('12000 P');
   });
   it('should check third tickets by price when alignment is "cheap"', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="cheap"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(screen.getByText(/15000 P/i)).toHaveTextContent('15000 P');
   });

   it('should check first tickets by speed when alignment is "speed"', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="speed"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(screen.getByText(/15000 P/i)).toHaveTextContent('15000 P');
   });
   it('should check second tickets by speed when alignment is "speed"', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="speed"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(screen.getByText(/12000 P/i)).toHaveTextContent('12000 P');
   });
   it('should check third tickets by speed when alignment is "speed"', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="speed"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(screen.getByText(/10000 P/i)).toHaveTextContent('10000 P');
   });

   it('should filters only tickets with 0 stops', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="cheap"
            transplants={{ ...baseTransplants, noneTransplants: true }}
         />,
      );
      expect(screen.getAllByText(/P/i)).toHaveLength(1);
   });

   it('should filters only tickets with 1 stop max', () => {
      render(
         <TicketsList
            list={mockTickets}
            alignment="cheap"
            transplants={{ ...baseTransplants, oneTransplants: true }}
         />,
      );
      const cards = screen.getAllByText(/P/i);
      expect(cards).toHaveLength(3);
   });
   it('should matches snapshot', () => {
      const { asFragment } = render(
         <TicketsList
            list={mockTickets}
            alignment="cheap"
            transplants={{ ...baseTransplants, all: true }}
         />,
      );
      expect(asFragment()).toMatchSnapshot();
   });
});
