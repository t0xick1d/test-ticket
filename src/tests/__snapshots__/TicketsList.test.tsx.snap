// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketsList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-4 containerList css-1ydw195-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="containerMainInf"
        >
          <p
            class="priceText"
          >
            10000 P
          </p>
          <p
            class="carrierText"
          >
            SU
          </p>
        </div>
        <ul
          class="containerSegments"
        >
          <li
            class="containerSegmentsItem"
          >
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                MOW - HKT
              </p>
              <p>
                10:30
              </p>
            </div>
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                В пути
              </p>
              <p>
                3ч 0м
              </p>
            </div>
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                0 пересадки
              </p>
              <div
                class="containerStops"
              />
            </div>
          </li>
        </ul>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="containerMainInf"
        >
          <p
            class="priceText"
          >
            12000 P
          </p>
          <p
            class="carrierText"
          >
            S7
          </p>
        </div>
        <ul
          class="containerSegments"
        >
          <li
            class="containerSegmentsItem"
          >
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                MOW - HKT
              </p>
              <p>
                11:00
              </p>
            </div>
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                В пути
              </p>
              <p>
                2ч 0м
              </p>
            </div>
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                1 пересадки
              </p>
              <div
                class="containerStops"
              >
                <p
                  class="pStops"
                >
                  DXB
                </p>
              </div>
            </div>
          </li>
        </ul>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="containerMainInf"
        >
          <p
            class="priceText"
          >
            15000 P
          </p>
          <p
            class="carrierText"
          >
            S7
          </p>
        </div>
        <ul
          class="containerSegments"
        >
          <li
            class="containerSegmentsItem"
          >
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                MOW - HKT
              </p>
              <p>
                11:00
              </p>
            </div>
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                В пути
              </p>
              <p>
                2ч 0м
              </p>
            </div>
            <div
              class="containerInfo"
            >
              <p
                class="secondaryText"
              >
                1 пересадки
              </p>
              <div
                class="containerStops"
              >
                <p
                  class="pStops"
                >
                  DXB
                </p>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TicketsList component should matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-4 containerList css-1ydw195-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="container"
        >
          <div
            class="containerMainInf"
          >
            <p
              class="priceText"
            >
              10000 P
            </p>
            <p
              class="carrierText"
            >
              SU
            </p>
          </div>
          <ul
            class="containerSegments"
          >
            <li
              class="containerSegmentsItem"
            >
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  MOW - HKT
                </p>
                <p>
                  10:30
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  В пути
                </p>
                <p>
                  3ч 0м
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  0 пересадки
                </p>
                <div
                  class="containerStops"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="container"
        >
          <div
            class="containerMainInf"
          >
            <p
              class="priceText"
            >
              12000 P
            </p>
            <p
              class="carrierText"
            >
              S7
            </p>
          </div>
          <ul
            class="containerSegments"
          >
            <li
              class="containerSegmentsItem"
            >
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  MOW - HKT
                </p>
                <p>
                  11:00
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  В пути
                </p>
                <p>
                  2ч 0м
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  1 пересадки
                </p>
                <div
                  class="containerStops"
                >
                  <p
                    class="pStops"
                  >
                    DXB
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="container"
        >
          <div
            class="containerMainInf"
          >
            <p
              class="priceText"
            >
              15000 P
            </p>
            <p
              class="carrierText"
            >
              S7
            </p>
          </div>
          <ul
            class="containerSegments"
          >
            <li
              class="containerSegmentsItem"
            >
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  MOW - HKT
                </p>
                <p>
                  11:00
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  В пути
                </p>
                <p>
                  2ч 0м
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  1 пересадки
                </p>
                <div
                  class="containerStops"
                >
                  <p
                    class="pStops"
                  >
                    DXB
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TicketsList component smatches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-4 containerList css-1ydw195-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="container"
        >
          <div
            class="containerMainInf"
          >
            <p
              class="priceText"
            >
              10000 P
            </p>
            <p
              class="carrierText"
            >
              SU
            </p>
          </div>
          <ul
            class="containerSegments"
          >
            <li
              class="containerSegmentsItem"
            >
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  MOW - HKT
                </p>
                <p>
                  10:30
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  В пути
                </p>
                <p>
                  3ч 0м
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  0 пересадки
                </p>
                <div
                  class="containerStops"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="container"
        >
          <div
            class="containerMainInf"
          >
            <p
              class="priceText"
            >
              12000 P
            </p>
            <p
              class="carrierText"
            >
              S7
            </p>
          </div>
          <ul
            class="containerSegments"
          >
            <li
              class="containerSegmentsItem"
            >
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  MOW - HKT
                </p>
                <p>
                  11:00
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  В пути
                </p>
                <p>
                  2ч 0м
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  1 пересадки
                </p>
                <div
                  class="containerStops"
                >
                  <p
                    class="pStops"
                  >
                    DXB
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="container"
        >
          <div
            class="containerMainInf"
          >
            <p
              class="priceText"
            >
              15000 P
            </p>
            <p
              class="carrierText"
            >
              S7
            </p>
          </div>
          <ul
            class="containerSegments"
          >
            <li
              class="containerSegmentsItem"
            >
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  MOW - HKT
                </p>
                <p>
                  11:00
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  В пути
                </p>
                <p>
                  2ч 0м
                </p>
              </div>
              <div
                class="containerInfo"
              >
                <p
                  class="secondaryText"
                >
                  1 пересадки
                </p>
                <div
                  class="containerStops"
                >
                  <p
                    class="pStops"
                  >
                    DXB
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
